### PR TITLE
ZB-1512 5: allow users to provide and access ethereum.logs table

### PR DESCRIPTION
**Pull request by mengtong-personal**
This PR adds support to allow user to provide and access `ethereum.logs` table in addition to `ethereum.blocks`.

A new struct `EthereumData` is provided to make it easier to add more tables in the future.
* It holds the customer provided data and handles the creation of the corresponding schemas and tables, thus making this struct the single place to provide data.
* It can be used both on the sourcedb and destdb. Typical use cases would be to provide custom data for sourcedb (to create the schemas and tables and populate data), and emptydata for destdb (to create the tables), although it is perfectly fine to provide data to destdb too.
* Support a new table would be to include it in the struct and model list. Everything else should just work.

The initialization of `ethereum.blocks` using the sql script is removed as it is substituted by the new struct.

The testing functions is modified to require a schema name (similar to the SourceSchema config) and to support multiple checkers.

### CODE REVIEW

File: testdata/init/ethereum_blocks.sql [hunk 1]
```diff
@@ -1,30 +1 @@
...
-  uncles VARCHAR(255) ARRAY,
-  num_of_transactions integer NOT NULL,
-  extra_data_raw TEXT,
-  extra_data TEXT,
-  process_time timestamp,
-  block_date DATE,
-  PRIMARY KEY (number)
-);
\ No newline at end of file
+CREATE SCHEMA IF NOT EXISTS ethereum;
```
Comment (bowenli86) [inline testdata/init/ethereum_blocks.sql]: if you look at zb'd db, there're 2 schemas - ethereum_mainnet, ethereum_holesky

Comment (mengtong-personal) [reply testdata/init/ethereum_blocks.sql]: updated

File: testexamples/ethereum_block_handler_test.go [hunk 2]
```diff
@@ -13,7 +13,13 @@ import (
 // destination table and custom table.
 func FindBlockHandlerString(blockNumber string, deps *utils.Deps) (bool, error) {
 	if blockNumber == "2" {
-		deps.DestinationDB.Create(&ethereum.Block{Number: 2})
+		// Exercise the code to read source logs table from the handler.
+		var logs []*ethereum.Log
+		deps.SourceDB.Table("ethereum.logs").Where("block_number = ?", 2).Find(&logs)
```
Comment (bowenli86) [inline testexamples/ethereum_block_handler_test.go:18]: "ethereum_mainnet.logs" or "ethereum_holesky.logs"?

File: testutils/ethereum_source_data.go [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+	data map[string]*EthereumSchemaData
+}
+
+func NewEthereumData() *EthereumData {
+	return &EthereumData{
+		data: make(map[string]*EthereumSchemaData),
+	}
+}
+
+func (d *EthereumData) AddSchemaData(schemaName string, blocks []*ethereum.Block, logs []*ethereum.Log) {
```
Comment (bowenli86) [inline testutils/ethereum_source_data.go:31]: what's the semantics? does all logs belong to a single block? or log to block is 1:1 mapping?

**Review by bowenli86 — COMMENTED** [review]
can you add formal documentations?

**Comment by mengtong-personal** [comment]
> can you add formal documentations?

what format of document are you referring to?

File: testexamples/ethereum_block_handler_test.go (hunk 2 above)
Comment (mengtong-personal) [reply testexamples/ethereum_block_handler_test.go:18]: what schema/table is available totally depends on the caller provided `EthereumData`. In this example I had the same data in all three schemas ethereum, ethereum_mainnet, ethereum_holesky so they all work, but I can update this one to be ethereum_holesky for better illustration.

File: testutils/ethereum_source_data.go (hunk 3 above)
Comment (mengtong-personal) [reply testutils/ethereum_source_data.go:31]: there is no enforcement of semantics by the test runner other than the primary keys defined in the model. Caller can provide any data they think appropriate for the testing. So, the logs do not have to belong to a single block and there isn't any mapping between log and block required by the test runner.

**Review by bowenli86 — APPROVED** [review]

**Pull request merged**